onload/cplane/main.go

```diff
@@ -0,0 +1,133 @@
+	cplaneServerParamsFilePath     = "/sys/module/onload/parameters/cplane_server_params"
+	critclFilePath                 = "/usr/bin/crictl"
+	crioPrefix                     = "cri-o://"
+	cplaneServerParamsFormatString = "exec %s /opt/onload/sbin/onload_cp_server -K"
+
+	timeoutTime = 5 * time.Second
+)
+
+var (
+	podName      string
+	podNamespace string
+
+	containerID string
+)
+
+func main() {
+
+	podName = os.Getenv("POD_NAME")
+	podNamespace = os.Getenv("POD_NAMESPACE")
+
+	cfg, err := config.GetConfig()
+	if err != nil {
+		glog.Fatalf("Failed to get client config (%v)", err)
+	}
+
+	c, err := client.New(cfg, client.Options{})
+	if err != nil {
+		glog.Fatalf("Failed to create client (%v)", err)
+	}
+	pod := &corev1.Pod{}
+
+	// The information about the pod may be out of date when it is received,
+	// so we enter a "while" loop to keep checking about the pod and container
+	// until they report as running.
+	for {
+		err = c.Get(context.Background(), client.ObjectKey{
+			Namespace: podNamespace,
+			Name:      podName,
+		}, pod)
+		if err != nil {
+			glog.Fatalf("Failed to find pod %q in namespace %q (%v)",
+				podName, podNamespace, err)
+		}
+
+		// We only expect a single container within the pod, so error out if a
+		// different number is found
+		if len(pod.Status.ContainerStatuses) != 1 {
+			glog.Fatalf("Expecting 1 containerStatus in pod %q, found %d",
+				podName, len(pod.Status.ContainerStatuses))
+		}
+		containerStatus := pod.Status.ContainerStatuses[0]
+
+		if !containerStatus.Ready {
+			glog.Warningf("Container is not ready. Will retry in %v",
+				timeoutTime)
+			time.Sleep(timeoutTime)
+			continue
+		}
+
+		if !*containerStatus.Started {
+			glog.Warningf("Container has not started. Will retry in %v",
+				timeoutTime)
+			time.Sleep(timeoutTime)
+			continue
+		}
+
+		if containerStatus.State.Running == nil {
+			glog.Warningf("Container is not running. Will retry in %v",
+				timeoutTime)
+			time.Sleep(timeoutTime)
+			continue
+		}
+
+		containerID = containerStatus.ContainerID
+
+		break
+	}
+
+	// The ContainerID reported has a "cri-o://" prefix before the actual ID.
+	// We need to remove this so that crictl is happy.
+	// If the prefix is not found, then we should be good to continue, but log
+	// the event anyway.
+	containerID, found := strings.CutPrefix(containerID, crioPrefix)
+	if !found {
+		glog.Warningf("Could not find %q as a prefix to the containerID %v",
+			crioPrefix, containerID)
+	}
+
+	err = os.WriteFile(
+		cplaneServerPathFilePath,
+		[]byte(critclFilePath),
+		0644,
+	)
+	if err != nil {
+		glog.Fatalf("Couldn't write cplane_server_path")
+	}
+
+	cplaneServerParams :=
+		fmt.Sprintf(cplaneServerParamsFormatString, containerID)
+
+	err = os.WriteFile(
+		cplaneServerParamsFilePath,
+		[]byte(cplaneServerParams),
+		0644,
+	)
+	if err != nil {
+		glog.Fatalf("Couldn't write cplane_server_params")
+	}
+
+	// We want this program to run indefinetly
```

Comment:
```suggestion
	// We want this program to run indefinitely
```

Reply:
Under what circumstances would the `containerID` change once found?

(Also, next line `select {}` is redundant?)

Reply:
I don't expect `containerID` to change, the problem is that if the program finishes then kubernetes will stop the DaemonSet and mark it as "Completed" which causes it to start the loop again.

We need a way to keep the pod alive indefinitely (or at least until the pod is removed). I could add something like `&& sleep infinitely` to the end of the command, but that would be ugly IMO.

`select {}` is used so that the program will block forever without eating cpu cycles.

Reply:
```suggestion
	// Block here indefinitely
```
Ah, I see it is a common hack, thanks. Wasn't how I assumed it'd behave. This would be clearer to me.